streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -833,8 +808,29 @@ private void commitOne(final AbstractTask task) {
         streamsMetrics.commitTimeSensor.record(computeLatency(), timerStartedMs);
     }
 
+    private void assignStandbyPartitions() {
+        final Collection<StandbyTask> running = standby.running();
+        final Map<TopicPartition, Long> checkpointedOffsets = new HashMap<>();
+        for (StandbyTask standbyTask : running) {
+            checkpointedOffsets.putAll(standbyTask.checkpointedOffsets());
+        }
+
+        final List<TopicPartition> assignment = new ArrayList<>(checkpointedOffsets.keySet());
+        restoreConsumer.assign(assignment);
+        for (final Map.Entry<TopicPartition, Long> entry : checkpointedOffsets.entrySet()) {
+            final TopicPartition partition = entry.getKey();
+            final long offset = entry.getValue();
+            if (offset >= 0) {
+                restoreConsumer.seek(partition, offset);
+            } else {
+                restoreConsumer.seekToBeginning(singleton(partition));
+            }
+        }
+        log.trace("{} assigned {} partitions to restore consumer for standby tasks {}", logPrefix, assignment, standby.runningTaskIds());
+    }
+
     private void maybeUpdateStandbyTasks(final long now) {
-        if (!standbyTasks.isEmpty()) {
+        if (state == State.RUNNING && standby.hasRunningTasks()) {
```

Comment:
I don't think so as it protects nothing. `state` is volatile anyway, so changes by another thread will be seen immediately. Further there is nothing to stop another thread from immediately changing the state to `PENDING_SHUTDOWN` immediately after this line is executed.